js/initcontext.go

```diff
@@ -110,7 +119,95 @@ func newBoundInitContext(base *InitContext, ctxPtr *context.Context, rt *goja.Ru
 		programs:          programs,
 		compatibilityMode: base.compatibilityMode,
 		logger:            base.logger,
+
+		shares:     base.shares,
+		sharesLock: base.sharesLock,
+	}
+}
+
+// NewShare ...
+// TODO rename
+func (i *InitContext) NewShare(ctx context.Context, name string, call goja.Callable) goja.Value {
+	i.sharesLock.Lock()
+	defer i.sharesLock.Unlock()
+	value, ok := i.shares[name]
+	rt := common.GetRuntime(ctx)
+	if !ok { //nolint:nestif
+		gojaValue, err := call(goja.Undefined())
+		if err != nil {
+			common.Throw(rt, err)
+		}
+		if gojaValue.ExportType().Kind() == reflect.Slice {
+			// fmt.Println("it's a slice")
+			var tmpArr []interface{}
+			if err = rt.ExportTo(gojaValue, &tmpArr); err != nil {
+				common.Throw(rt, err)
+			}
+
+			arr := make([][]byte, len(tmpArr))
+			for i := range arr {
+				arr[i], err = json.Marshal(tmpArr[i])
+				if err != nil {
+					common.Throw(rt, err)
+				}
+			}
+			value = sharedArray{arr: arr}
+		} else {
+			// fmt.Println("it's an object")
+			value = shared{gojaValue.ToObject(rt)}
+		}
+		i.shares[name] = value
 	}
+	switch value.(type) {
+	case sharedArray:
+		// fmt.Println("wrapping array/slice")
+		// TODO cache this
+		cal, err := rt.RunString(`(function(val) {
+	var arrayHandler = {
+		get: function(target, property, receiver) {
+			// console.log("accessing ", property)
+			switch (property){
+			case "length":
+				return target.length()
+			case Symbol.iterator:
+				return function() {return target.iterator()}
```

Comment:
This unfortunately can't be done through [goja.ProxyTrapConifg](https://pkg.go.dev/github.com/dop251/goja#ProxyTrapConfig) So if we want to have support for [for-of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) We need to use js proxy (we can use both, but I don't think that will help with anything :(

Reply:
Hmm why can't we use `goja.ProxyTrapConifg`? Because of `Symbol.iterator`? If so, can't we just get the goja definition of the iterator symbol from the runtime and compare against that in the `Get` property of the `goja.ProxyTrapConifg`?

Reply:
Yes, 
I am gonna copy you the relevant part(which is most of it) of documentation that I linked ;) 

> Note that the Proxy may not have Symbol properties when using this as a handler because property keys are passed as strings. get() and getOwnPropertyDescriptor() for Symbol properties will always return undefined; has() and deleteProperty() for Symbol properties will always return false; set() and defineProperty() for Symbol properties will throw a TypeError. If you need Symbol properties implement the handler in JavaScript.

Reply:
This will probably take some breaking of the API to be fixed and also the standard Symbols will need to be exported by goja .. which they currently aren't. 

I doubt this will actually make much of a performance difference to be honest ... But it will be kind of nice to not have to execute a script to wrap it I guess

Reply:
Although i guess it is worth an issue to ask why the name wasn't a goja.Value to begin with instead of string

Reply:
We can't access the symbol from the Go side, but can't we get it from the JS side by just getting `Symbol.iterator` from the runtime, even now?

Reply:
Ah, nevermind, somehow missed the first of your 3 comments...